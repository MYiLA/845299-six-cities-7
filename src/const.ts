export enum AppRoute {
  MAIN = 'main',
  LOGIN = 'login',
  FAVORITES = 'favorites',
  OFFER = 'offer',
  DEFAULT_CITY = 'Paris'
};

export enum APIRoute {
  HOTELS = '/hotels',
  FAVORITE = '/favorite',
  COMMENTS = '/comments',
  LOGIN = '/login',
  LOGOUT = '/logout',
};

export enum CardType {
  FAVORITES = 'favorites',
  CITIES = 'cities',
  NEAR_PLACES = 'near-places'
}

export enum AuthorizationStatus {
  AUTH = 'AUTH',
  NO_AUTH = 'NO_AUTH',
  UNKNOWN = 'UNKNOWN',
};

export const Cities = [
  {
    location: {
      latitude: 48.85,
      longitude: 2.35,
      zoom: 10,
    },
    name: 'Paris',
  },
  {
    location: {
      latitude: 50.9375,
      longitude: 6.9603,
      zoom: 10,
    },
    name: 'Cologne',
  },
  {
    location: {
      latitude: 50.8503,
      longitude: 4.3517,
      zoom: 10,
    },
    name: 'Brussels',
  },
  {
    location: {
      latitude: 52.370216,
      longitude: 4.895168,
      zoom: 10,
    },
    name: 'Amsterdam',
  },
  {
    location: {
      latitude: 53.5511,
      longitude: 9.9937,
      zoom: 10,
    },
    name: 'Hamburg',
  },
  {
    location: {
      latitude: 51.2277,
      longitude: 6.7735,
      zoom: 10,
    },
    name: 'Dusseldorf',
  },
];

export const RegularExpression = {
  EMAIL: /^[^\s@]+@[^\s@]+\.[^\s@]+$/,
  ONE_SIMBOL: /\S/,
}
