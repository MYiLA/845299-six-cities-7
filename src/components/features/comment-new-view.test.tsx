import { fireEvent, render } from '@testing-library/react';
import { SNAP_DESC } from '../../const';
import CommentNewView from './comment-new-view';

let onSubmit = () => null;
let onChange = () => null;
let setComment = () => null;
let setRating = () => null;
const rating = 3;
const comment = 'Этот комментарий слишком короткий';
const commentNew = 'Комментарий нормальной длинны, чтобы можно было отправить и проверить отправку формы';
const isFormDisabled = true;
const isSubmitDisabled = true;

describe('поведение компонента bookmark-view', () => {
  beforeEach(() => {
    onSubmit = jest.fn();
    onChange = jest.fn();
    setComment = jest.fn();
    setRating = jest.fn();
  });

  it('отрисовывает рейтинг', () => {
    const { container } = render(
      <CommentNewView
        onSubmit={onSubmit}
        onChange={onChange}
        setComment={setComment}
        setRating={setRating}
        rating={rating}
        comment={comment}
        isFormDisabled={isFormDisabled}
        isSubmitDisabled={isSubmitDisabled}
      />,
    );
    const inputStar = container.querySelector('[id=\'3-stars\']') as HTMLInputElement;
    expect(inputStar.checked).toBe(true);
  });

  it('отрисовывает комментарий', () => {
    const { container } = render(
      <CommentNewView
        onSubmit={onSubmit}
        onChange={onChange}
        setComment={setComment}
        setRating={setRating}
        rating={rating}
        comment={comment}
        isFormDisabled={isFormDisabled}
        isSubmitDisabled={isSubmitDisabled}
      />,
    );

    expect(container.querySelector('.reviews__textarea.form__textarea')?.textContent).toBe(comment);
  });

  it('блокирует форму', () => {
    const { container } = render(
      <CommentNewView
        onSubmit={onSubmit}
        onChange={onChange}
        setComment={setComment}
        setRating={setRating}
        rating={rating}
        comment={comment}
        isFormDisabled={isFormDisabled}
        isSubmitDisabled={isSubmitDisabled}
      />,
    );
    const formElement = container.querySelector('.form__textarea') as HTMLTextAreaElement;
    const inputElement = container.querySelector('.form__rating-input') as HTMLInputElement;
    expect(formElement.disabled).toBe(true);
    expect(inputElement.disabled).toBe(true);
  });

  it('блокирует кнопку отправки', () => {
    const { container } = render(
      <CommentNewView
        onSubmit={onSubmit}
        onChange={onChange}
        setComment={setComment}
        setRating={setRating}
        rating={rating}
        comment={comment}
        isFormDisabled={isFormDisabled}
        isSubmitDisabled={isSubmitDisabled}
      />,
    );

    const submitElement = container.querySelector('.form__submit') as HTMLButtonElement;
    expect(submitElement.disabled).toBe(true);
  });

  it('обрабатывает изменения в комментарии', () => {
    const { container } = render(
      <CommentNewView
        onSubmit={onSubmit}
        onChange={onChange}
        setComment={setComment}
        setRating={setRating}
        rating={rating}
        comment={comment}
        isFormDisabled={false}
        isSubmitDisabled={isSubmitDisabled}
      />,
    );

    const textareaElement = container.querySelector('.form__textarea') as HTMLTextAreaElement;
    fireEvent.change(textareaElement, { target: { value: commentNew } });
    expect(onChange).toBeCalled();
    expect(setComment).toBeCalled();
  });

  it('обрабатывает изменения в рейтинге', () => {
    const { container } = render(
      <CommentNewView
        onSubmit={onSubmit}
        onChange={onChange}
        setComment={setComment}
        setRating={setRating}
        rating={rating}
        comment={comment}
        isFormDisabled={false}
        isSubmitDisabled={isSubmitDisabled}
      />,
    );
    const inputStar = container.querySelector('[id=\'2-stars\']') as HTMLInputElement;
    fireEvent.click(inputStar);
    expect(onChange).toBeCalled();
    expect(setRating).toBeCalled();
  });

  it('обрабатывает отправку отзыва', () => {
    const { container } = render(
      <CommentNewView
        onSubmit={onSubmit}
        onChange={onChange}
        setComment={setComment}
        setRating={setRating}
        rating={rating}
        comment={comment}
        isFormDisabled={isFormDisabled}
        isSubmitDisabled={false}
      />,
    );

    const submitElement = container.querySelector('.form__submit') as HTMLButtonElement;
    fireEvent.submit(submitElement);
    expect(onSubmit).toBeCalled();
  });

  it(SNAP_DESC, () => {
    const { container } = render(
      <CommentNewView
        onSubmit={onSubmit}
        onChange={onChange}
        setComment={setComment}
        setRating={setRating}
        rating={rating}
        comment={comment}
        isFormDisabled={isFormDisabled}
        isSubmitDisabled={false}
      />,
    );
    expect(container).toMatchSnapshot();
  });
});
