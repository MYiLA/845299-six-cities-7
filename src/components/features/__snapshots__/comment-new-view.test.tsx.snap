// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`поведение компонента bookmark-view нет изменений в разметке, по сравнению с прошлым тестом 1`] = `
<div>
  <form
    action="#"
    class="reviews__form form"
    method="post"
  >
    <label
      class="reviews__label form__label"
      for="review"
    >
      Your review
    </label>
    <div
      class="reviews__rating-form form__rating"
    >
      <input
        class="form__rating-input visually-hidden"
        disabled=""
        id="5-stars"
        name="rating"
        type="radio"
        value="5"
      />
      <label
        class="reviews__rating-label form__rating-label"
        for="5-stars"
        title="perfect"
      >
        <svg
          class="form__star-image"
          height="33"
          width="37"
        >
          <use
            xlink:href="#icon-star"
          />
        </svg>
      </label>
      <input
        class="form__rating-input visually-hidden"
        disabled=""
        id="4-stars"
        name="rating"
        type="radio"
        value="4"
      />
      <label
        class="reviews__rating-label form__rating-label"
        for="4-stars"
        title="perfect"
      >
        <svg
          class="form__star-image"
          height="33"
          width="37"
        >
          <use
            xlink:href="#icon-star"
          />
        </svg>
      </label>
      <input
        checked=""
        class="form__rating-input visually-hidden"
        disabled=""
        id="3-stars"
        name="rating"
        type="radio"
        value="3"
      />
      <label
        class="reviews__rating-label form__rating-label"
        for="3-stars"
        title="perfect"
      >
        <svg
          class="form__star-image"
          height="33"
          width="37"
        >
          <use
            xlink:href="#icon-star"
          />
        </svg>
      </label>
      <input
        class="form__rating-input visually-hidden"
        disabled=""
        id="2-stars"
        name="rating"
        type="radio"
        value="2"
      />
      <label
        class="reviews__rating-label form__rating-label"
        for="2-stars"
        title="perfect"
      >
        <svg
          class="form__star-image"
          height="33"
          width="37"
        >
          <use
            xlink:href="#icon-star"
          />
        </svg>
      </label>
      <input
        class="form__rating-input visually-hidden"
        disabled=""
        id="1-stars"
        name="rating"
        type="radio"
        value="1"
      />
      <label
        class="reviews__rating-label form__rating-label"
        for="1-stars"
        title="perfect"
      >
        <svg
          class="form__star-image"
          height="33"
          width="37"
        >
          <use
            xlink:href="#icon-star"
          />
        </svg>
      </label>
    </div>
    <textarea
      class="reviews__textarea form__textarea"
      disabled=""
      id="review"
      name="review"
      placeholder="Tell how was your stay, what you like and what can be improved"
    >
      Этот комментарий слишком короткий
    </textarea>
    <div
      class="reviews__button-wrapper"
    >
      <p
        class="reviews__help"
      >
        To submit review please make sure to set
         
        <span
          class="reviews__star"
        >
          rating
        </span>
         
        and describe your stay with at least
         
        <b
          class="reviews__text-amount"
        >
          50 characters
        </b>
        .
      </p>
      <button
        class="reviews__submit form__submit button"
        type="submit"
      >
        Submit
      </button>
    </div>
  </form>
</div>
`;
